### PR TITLE
Drawer

### PR DESCRIPTION
**Pull request by daileytj**
Fixes PXBLUE-1092, 1093, 1259.

Changes proposed in this Pull Request:
- Add drawer component

### CODE REVIEW

File: components/src/core/drawer/drawer.component.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            </div>
+        </div>
+    `,
+    styleUrls: ['./drawer.component.scss'],
+})
+export class DrawerComponent implements OnInit {
+    @Input() variant: VariantType;
+    @Input() variantDrawerHandler: boolean;
+    @Output() drawerOpenChange: EventEmitter<boolean> = new EventEmitter();
+    drawerOpen: boolean;
```
Comment (joebochill) [inline components/src/core/drawer/drawer.component.ts:36]: Where does drawerOpen actually get used to change the appearance of the drawer from navigation rail to fully opened (for persistent)?

**Comment by joebochill** [comment]
Comments above are just based on looking at the code. I'm going through now and attempting to use it in a sample project.

File: components/src/core/drawer/drawer.component.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                <ng-content select="pxb-drawer-footer"></ng-content>
+            </div>
+        </div>
+    `,
+    styleUrls: ['./drawer.component.scss'],
+})
+export class DrawerComponent implements OnInit {
+    @Input() variant: VariantType;
+    @Input() variantDrawerHandler: boolean;
+    @Output() drawerOpenChange: EventEmitter<boolean> = new EventEmitter();
```
[the remark this answers is not in the document]
Comment (huayunh) [reply components/src/core/drawer/drawer.component.ts:35]: Or should we allow both?

**Comment by joebochill** [comment]
Need to add new components checks in the `buildTest.sh` file

**Comment by joebochill** [comment]
Ideally, I would like it to work exactly like the mat-sidenav with two-way data binding on `drawerOpen` like in this example:
https://stackblitz.com/angular/gqednpxjnde?file=src%2Fapp%2Fsidenav-open-close-example.html

**Comment by joebochill** [comment]
I tried publishing this as a beta package for testing @2.0.0-beta.5, but it throws a bunch of compilation errors and doesn't run when you pull it into a project:
image

**Comment by joebochill** [comment]
A couple of comments about the README too:
- main image is really wide...should it have a max width?
- missing the import instructions in the Usage section

File: components/src/core/drawer/drawer-body/drawer-body.component.scss [hunk 3]
```diff
@@ -0,0 +1,11 @@
+.pxb-drawer-body {
+    width: 100%;
+}
+
+.pxb-drawer-nested-nav-item {
+    .pxb-info-list-item .mat-list-item-content {
+        padding-left: 24px;
+        font-size: 12px;
```
[the remark this answers is not in the document]
Comment (daileytj) [reply components/src/core/drawer/drawer-body/drawer-body.component.scss]: removed font-size styles

File: components/src/core/drawer/drawer-nav-item/drawer-nav-item.component.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                [matRippleDisabled]="!ripple"
+                style="display:flex;"
+            >
+                <div class="pxb-drawer-nav-item-icon-wrapper" icon>
+                    <ng-content select="[icon]"></ng-content>
+                </div>
+                <div title>{{ title }}</div>
+                <div subtitle>{{ subtitle }}</div>
+                <div rightContent *ngIf="hasChildren && expandIcon && !showNestedNavItems">
+                    <mat-icon
```
[the remark this answers is not in the document]
Comment (daileytj) [reply components/src/core/drawer/drawer-nav-item/drawer-nav-item.component.ts:82]: The "right content" of the pxb-info-list-item where the collapse/expand icons would be rendered uses content projection so I as having a hard time projecting content into projected content. I don't think we're able to do that with Angular, I'm pretty sure we can only pass projected content one layer deep... but I could be wrong. We can look into alternative solutions for this though.

File: components/src/core/drawer/drawer.component.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                <ng-content select="pxb-drawer-body"></ng-content>
+                <ng-content select="pxb-drawer-footer"></ng-content>
+            </div>
+        </div>
+    `,
+    styleUrls: ['./drawer.component.scss'],
+})
+export class DrawerComponent implements OnInit {
+    @Input() variant: VariantType;
+    @Input() variantDrawerHandler: boolean;
```
[the remark this answers is not in the document]
Comment (daileytj) [reply components/src/core/drawer/drawer.component.ts:34]: Agreed, I don't love my solution for this and I actually have a hard time explaining what it does. We can look into refactoring this functionality.

File: components/src/core/drawer/drawer.component.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+@Component({
+    selector: 'pxb-drawer',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+    encapsulation: ViewEncapsulation.None,
+    template: `
+        <div class="pxb-drawer">
+            <!-- Drawer is responsible for managing the styles between the 4 subsections -->
+            <ng-content select="pxb-drawer-header"></ng-content>
+            <div (mouseenter)="openDrawer()" (mouseleave)="closeDrawer()">
```
[the remark this answers is not in the document]
Comment (daileytj) [reply components/src/core/drawer/drawer.component.ts]: done! Renamed these methods.

File: demos/showcase/src/app/drawer.component.scss [hunk 7]
```diff
@@ -0,0 +1,46 @@
+@import '~@pxblue/colors/palette.scss';
+
```
[the remark this answers is not in the document]
Comment (daileytj) [reply demos/showcase/src/app/drawer.component.scss:2]: I made a PR for themes, once this is merged and a new beta package is released I can remove the styles.

File: demos/showcase/src/app/app.component.html [hunk 8]
```diff
@@ -1,263 +1,219 @@
...
+                <mat-icon emptyIcon>devices</mat-icon>
+                <button mat-raised-button color="primary" actions>
+                    <mat-icon>add_circle</mat-icon>
+                    ADD DEVICE
+                </button>
+            </pxb-empty-state>
+        </mat-card>
+
+        <mat-card class="container-card" style="margin-top: 10px; margin-bottom: 24px;">
+            <pxb-drawer-header title="im the title" class="test-background-image"></pxb-drawer-header>
```
[the remark this answers is not in the document]
Comment (daileytj) [reply demos/showcase/src/app/app.component.html]: removed the drawer header examples

**Comment by joebochill** [comment]
Here are a few observations from playing with the drawer in the showcase (we may want to save some of these for an enhancements push, but wanted to make sure I captured them):

### Drawer Components
* Default width should match react
* Collapsed persistent drawer should be 56px wide
* Should be an animation on the persistent drawer on hover (same as temporary when it opens)
* Hover overlay on DrawerNavItem should be slightly transparent (if you hover over an item and select it, you can’t see that the background color has changed)
* Submenu item expander icons should also animate (rotate) like the top-level ones
* Footer should be pinned to the bottom of the Drawer (not stuck to the bottom of body content)
* Header background image should not disappear when collapsed
* User should not have to set the spacing/margins in the drawer header icon for keyline alignment
* Drawer Subheader content overflows the drawer (horizontal) when expanded…looks like an issue with box-sizing
    * Our theme should be setting * { box-sizing: border-box }
* Why does drawerSubheader have a fixed height?
* Nested items are not working below level 2


### Showcase
* Update showcase with a top toolbar with the menu icon
* Add drawer mode toggler into toolbar (dropdown or toggle button)

File: components/src/core/drawer/drawer-nav-item/drawer-nav-item.component.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (joebochill) [reply components/src/core/drawer/drawer-nav-item/drawer-nav-item.component.ts:82]: We can leave it as is for now, and maybe log an issue for an enhancement to support alternative icons

**Comment by emclaug2** [comment]
In addition to Joe's comments there's a few observations that I have as well: 

- DrawerBodyModule doesn't need to import MatDividerModule.
- The arrangement of the modules is a little confusing to me; it's unclear why DrawerNavGroupComponent would be a part of the DrawerBodyModule, but DrawerNavItemComponent has its own module.  I think they should either be grouped into the body module or DrawerNavGroup should be separated into its own module. 
- The DrawerFooter & DrawerSubheader visibility style should be configurable through css classes so users can override the decision to hide the footer when closed.
- DrawerLayoutComponent's DrawerService should be private; whenever it's public it makes me think it's used within the template or in the unit tests. 
- When expanding the NestedItems, the text overlaps the NavItem below it while it adjusts its height.
- MenuIcon needs to be centered when the Drawer is skinny/closed. 
- NestedNavItems needs to be collapsed when the Drawer is closed.
- There should be a small delay before the Drawer is opened through a hover event; right now it's instant. 
- DrawerBody doesn't need the `drawerOpen` prop or the service.

**Comment by joebochill** [comment]
I'm going to approve this, and we can handle most of these update requests in PXBLUE-1383.

**Pull request merged**